analytics/src/main/java/com/segment/analytics/internal/AnalyticsClient.java

```diff
@@ -358,4 +394,86 @@ private static boolean is5xx(int status) {
       return status >= 500 && status < 600;
     }
   }
+
+  public static class BatchUtility {
+    /**
+     * Method to create a batch considering threshold of 500Kb for entire batch
+     *
+     * @param list
+     * @return batch object
+     */
+    public static Batch createBatch(
+        LinkedList<Message> list, Map<String, Integer> messageIdSizeMap, int contextSize) {
+      List<Message> messagesForBatch = new ArrayList<>();
+      Batch batch = null;
+
+      int currentBatchSize = 0;
+
+      // since we are handling max size of 32 kbs when enqueuing msg, one single msg will not be
+      // above limit
+      while (list.size() > 0 && currentBatchSize <= BATCH_MAX_SIZE) {
+        Message msg = list.peek();
+        int msgSize = messageIdSizeMap.get(msg.messageId());
+        int defaultBatchSizeSoFar = getBatchDefaultSize(contextSize, messagesForBatch.size() + 1);
```

Comment:
why are we computing this value for every message? it seems independent from the message

Reply:
@wenxi-zeng  the problem is the default batch size wont always be the same, the number of msgs determine the number of commas which is an extra character we have to account for

Reply:
if it's only for commas, I believe you can use this equation: 
`default batch size = base batch size + number of commas`
where 
`number of commas = number of messages - 1`
since each `char` occupies 1 byte

from another perspective we are adding 1 commas/1 byte to `defaultBatchSizeSoFar` for each message. we can just compute the base before the loop, and increased by 1 each loop. 

but nevermind, the new approach you proposed can get rid of this method, so don't worry about this.